rules/kuiper/sink.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/emqx/kuiper/common"
+	"github.com/emqx/kuiper/xstream/api"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/pkg/messaging/nats"
+	"github.com/mainflux/senml"
+)
+
+const baseName = "rules_engine"
+
+type mainfluxConfig struct {
+	Host     string `json:"host"`
+	Port     string `json:"port"`
+	Channel  string `json:"channel"`
+	Subtopic string `json:"subtopic"`
+}
+
+type mainfluxSink struct {
+	cfg *mainfluxConfig
+	pub nats.Publisher
+}
+
+// Configure sink with properties from rule action definition
+func (ms *mainfluxSink) Configure(props map[string]interface{}) error {
+	cfg := &mainfluxConfig{}
+
+	if err := common.MapToStruct(props, cfg); err != nil {
+		return fmt.Errorf("Read properties %v fail with error: %v", props, err)
+	}
+	if cfg.Host == "" {
+		return fmt.Errorf("property Host is required")
+	}
+	if cfg.Port == "" {
+		return fmt.Errorf("property Port is required")
+	}
+	if cfg.Channel == "" {
+		return fmt.Errorf("property Channel is required")
+	}
+
+	ms.cfg = cfg
+
+	return nil
+}
+
+func (ms *mainfluxSink) Open(ctx api.StreamContext) (err error) {
+	logger := ctx.GetLogger()
+	logger.Debug("Opening mainflux sink")
+
+	addr := fmt.Sprintf("%s:%s/", ms.cfg.Host, ms.cfg.Port)
+	pub, err := nats.NewPublisher(addr)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to nats at address %s with error: %v", addr, err)
+	}
+	ms.pub = pub
+
+	return
+}
+
+// Collect publishes messages transferred to sink to nats
+func (ms *mainfluxSink) Collect(ctx api.StreamContext, item interface{}) error {
+	logger := ctx.GetLogger()
+	logger.Debugf("mainflux sink receive %v", item)
+
+	var msg messaging.Message
+	msg.Channel = ms.cfg.Channel
+	msg.Subtopic = ms.cfg.Subtopic
+	msg.Created = time.Now().Unix()
+
+	itemBytes, ok := item.([]byte)
+	if !ok {
+		logger.Debug("mainflux sink received non byte data")
+	}
+	var rec []senml.Record
+	if err := json.Unmarshal(itemBytes, &rec); err != nil {
+		return fmt.Errorf("Failed to unmarshal %v to senml", item)
+	}
+	if rec[0].BaseName == "" {
```

Comment:
do we need to check the array before accesing its elements?

Reply:
Kuiper will call Collect only when there are messages to process, so there will always be at least 1 message. Actually, kuiper always sends one single message per senml record.

Reply:
I'd add a `len()` check in any case

Reply:
It's impossible that this array be empty. Should i proceed with the check?

Reply:
I'd add it yes. If one day you update Kuiper and they changed the behaviour of this function you will have a segmentation fault....